File: python/src/SimpleTokenBuilder.py
import random
import time
import os
import sys
from collections import OrderedDict
sys.path.append(os.path.join(os.path.dirname(__file__), '../src'))
from AccessToken import *


ROLE_ATTENDEE = 1
ROLE_PUBLISHER = 2
ROLE_SUBSCRIBER = 3
ROLE_ADMIN = 4

AttendeePrivileges = OrderedDict ([
	(kJoinChannel, 0),
	(kPublishAudioStream, 0),
	(kPublishVideoStream, 0),
	(kPublishDataStream, 0)
])

PublisherPrivileges = OrderedDict ([
	(kJoinChannel, 0),
	(kPublishAudioStream, 0), 
	(kPublishVideoStream, 0),
	(kPublishDataStream, 0),
	(kPublishAudiocdn, 0),
	(kPublishVideoCdn, 0),
	(kInvitePublishAudioStream, 0),
	(kInvitePublishVideoStream, 0),
	(kAdministrateChannel, 0)
])

SubscriberPrivileges = OrderedDict ([
	(kJoinChannel, 0),
	(kRequestPublishAudioStream, 0),
	(kRequestPublishVideoStream, 0),
	(kRequestPublishDataStream, 0)
])

AdminPrivileges = OrderedDict ([
	(kJoinChannel, 0),
	(kPublishAudioStream, 0), 
	(kPublishVideoStream, 0),
	(kPublishDataStream, 0),
	(kAdministrateChannel, 0)
])

RolePrivileges = OrderedDict ([
	(ROLE_ATTENDEE, AttendeePrivileges),
	(ROLE_PUBLISHER, PublisherPrivileges),
	(ROLE_SUBSCRIBER, SubscriberPrivileges), 
	(ROLE_ADMIN, AdminPrivileges)
])

class SimpleTokenBuilder:

    def __init__(self, appID, appCertificate, channelName, uid):   
        self.token = AccessToken(appID, appCertificate, channelName, uid)
        random.seed(time.time())
        self.token.ts = int(time.time()) + 24 * 3600
        self.token.salt = random.randint(1, 99999999)
        self.token.messages = {}

    def initPriviliges(self, role) :
	    self.token.messages = RolePrivileges[role]

    def setPrivilege(self, privilege, timeoutFromNow) :
	    self.token.messages[privilege] = timeoutFromNow

    def removePrivilege(self, privilege) :
	    self.token.messages.pop(privilege)

    def buildToken(self):
        return self.token.build()

File: nodejs/src/SimpleTokenBuilder.js
var AccessToken = require('../src/AccessToken').AccessToken;
var Priviledges = require('../src/AccessToken').priviledges;
var randomInt = Math.floor(Math.random() * 0xFFFFFFFF);

var Role = {
    kRoleAttendee: 0,  // for communication
    kRolePublisher: 1, // for live broadcast
    kRoleSubscriber: 2,  // for live broadcast
    kRoleAdmin: 101
};

var attendeePrivileges = {
    [Priviledges.kJoinChannel]: 0,
    [Priviledges.kPublishAudioStream]: 0,
    [Priviledges.kPublishVideoStream]: 0,
    [Priviledges.kPublishDataStream]: 0
};


var publisherPrivileges = {
    [Priviledges.kJoinChannel]: 0,
    [Priviledges.kPublishAudioStream]: 0,
    [Priviledges.kPublishVideoStream]: 0,
    [Priviledges.kPublishDataStream]: 0,
    [Priviledges.kPublishAudiocdn]: 0,
    [Priviledges.kPublishVideoCdn]: 0,
    [Priviledges.kInvitePublishAudioStream]: 0,
    [Priviledges.kInvitePublishVideoStream]: 0,
    [Priviledges.kInvitePublishDataStream]: 0
};

var subscriberPrivileges = {
    [Priviledges.kJoinChannel] : 0,
    [Priviledges.kRequestPublishAudioStream] : 0,
    [Priviledges.kRequestPublishVideoStream] : 0,
    [Priviledges.kRequestPublishDataStream] : 0
};

var adminPrivileges = {
    [Priviledges.kJoinChannel] : 0,
    [Priviledges.kPublishAudioStream] : 0,
    [Priviledges.kPublishVideoStream] : 0,
    [Priviledges.kPublishDataStream] : 0,
    [Priviledges.kAdministrateChannel]: 0
};

var RolePrivileges = {
	[Role.kRoleAttendee]: attendeePrivileges,
    [Role.kRolePublisher]: publisherPrivileges,
    [Role.kRoleSubscriber]: subscriberPrivileges,
    [Role.kRoleAdmin]: adminPrivileges
}

var SimpleTokenBuilder = function (appID, appCertificate, channelName, uid) {
    let builder = this;
    this.key = new AccessToken(appID, appCertificate, channelName, uid)
    this.key.salt = randomInt;
    this.key.ts = new Date().getTime() / 1000 + 24 * 3600;
    this.key.messages = {};

    this.buildToken = function () {
        return builder.key.build()
    }

    this.initPriviliges = function (role) {
        let rolePri = RolePrivileges[role];
        builder.key.messages = JSON.parse(JSON.stringify(rolePri));
    }
    this.setPrivilege = function (privilege, timeoutFromNow) {
        builder.key.messages[privilege] = timeoutFromNow
    }
    this.removePrivilege = function (privilege) {
        delete builder.key.messages[privilege]
    }
};

module.exports.SimpleTokenBuilder = SimpleTokenBuilder;
module.exports.Role = Role;